From 0735edf47213f70460602b9f25164f4d66576aa3 Mon Sep 17 00:00:00 2001
From: Danila Tikhonov <danila@jiaxyga.com>
Date: Wed, 10 Apr 2024 20:46:50 +0300
Subject: [PATCH 084/160] media: qcom: camss: Add SM7150 resources

This commit describes the hardware layout for the SM7150 for the
following hardware blocks:

- 4 x CSI PHY
- 2 x CSID
- 2 x VFE, 4 RDI per VFE
- 1 x VFE Lite, 4 RDI per VFE

Signed-off-by: Danila Tikhonov <danila@jiaxyga.com>
---
 drivers/media/platform/qcom/camss/camss.c | 203 ++++++++++++++++++++++
 1 file changed, 203 insertions(+)

diff --git a/drivers/media/platform/qcom/camss/camss.c b/drivers/media/platform/qcom/camss/camss.c
index 2123f6388e3d..6f36a28fbf6b 100644
--- a/drivers/media/platform/qcom/camss/camss.c
+++ b/drivers/media/platform/qcom/camss/camss.c
@@ -1950,6 +1950,194 @@ static const struct resources_icc icc_res_sm6350[] = {
 	},
 };
 
+static const struct camss_subdev_resources csiphy_res_7150[] = {
+	/* CSIPHY0 */
+	{
+		.regulators = {},
+		.clock = { "cphy_rx_src", "csiphy0", "csiphy0_timer",
+			   "csiphy0_timer_src" },
+		.clock_rate = { { 384000000, 400000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000 } },
+		.reg = { "csiphy0" },
+		.interrupt = { "csiphy0" },
+		.csiphy = {
+			.hw_ops = &csiphy_ops_3ph_1_0,
+			.formats = &csiphy_formats_sdm845
+		}
+	},
+	/* CSIPHY1 */
+	{
+		.regulators = {},
+		.clock = { "cphy_rx_src", "csiphy1", "csiphy1_timer",
+			   "csiphy1_timer_src" },
+		.clock_rate = { { 384000000, 400000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000 } },
+		.reg = { "csiphy1" },
+		.interrupt = { "csiphy1" },
+		.csiphy = {
+			.hw_ops = &csiphy_ops_3ph_1_0,
+			.formats = &csiphy_formats_sdm845
+		}
+	},
+	/* CSIPHY2 */
+	{
+		.regulators = {},
+		.clock = { "cphy_rx_src", "csiphy2", "csiphy2_timer",
+			   "csiphy2_timer_src" },
+		.clock_rate = { { 384000000, 400000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000 } },
+		.reg = { "csiphy2" },
+		.interrupt = { "csiphy2" },
+		.csiphy = {
+			.hw_ops = &csiphy_ops_3ph_1_0,
+			.formats = &csiphy_formats_sdm845
+		}
+	},
+	/* CSIPHY3 */
+	{
+		.regulators = {},
+		.clock = { "cphy_rx_src", "csiphy3", "csiphy3_timer",
+			   "csiphy3_timer_src" },
+		.clock_rate = { { 384000000, 400000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000 } },
+		.reg = { "csiphy3" },
+		.interrupt = { "csiphy3" },
+		.csiphy = {
+			.hw_ops = &csiphy_ops_3ph_1_0,
+			.formats = &csiphy_formats_sdm845
+		}
+	}
+};
+
+static const struct camss_subdev_resources csid_res_7150[] = {
+	/* CSID0 */
+	{
+		.regulators = { "vdda-phy", "vdda-pll" },
+		.clock = { "cphy_rx_src", "vfe0_axi", "vfe0", "vfe0_src",
+			   "vfe0_cphy_rx", "vfe0_csid", "vfe0_csid_src", },
+		.clock_rate = { { 0 },
+				{ 0 },
+				{ 0 },
+				{ 380000000, 510000000, 637000000, 760000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000, 384000000, 400000000 } },
+		.reg = { "csid0" },
+		.interrupt = { "csid0" },
+		.csid = {
+			.hw_ops = &csid_ops_gen2,
+			.parent_dev_ops = &vfe_parent_dev_ops,
+			.formats = &csid_formats_gen2
+		}
+	},
+	/* CSID1 */
+	{
+		.regulators = { "vdda-phy", "vdda-pll" },
+		.clock = { "cphy_rx_src", "vfe0_axi", "vfe1", "vfe1_src",
+			   "vfe1_cphy_rx", "vfe1_csid", "vfe0_csid_src", },
+		.clock_rate = { { 0 },
+				{ 0 },
+				{ 0 },
+				{ 380000000, 510000000, 637000000, 760000000 },
+				{ 0 },
+				{ 0 },
+				{ 300000000, 384000000, 400000000 } },
+		.reg = { "csid1" },
+		.interrupt = { "csid1" },
+		.csid = {
+			.hw_ops = &csid_ops_gen2,
+			.parent_dev_ops = &vfe_parent_dev_ops,
+			.formats = &csid_formats_gen2
+		}
+	}
+};
+
+static const struct camss_subdev_resources vfe_res_7150[] = {
+	/* VFE0 */
+	{
+		.regulators = {},
+		.clock = { "vfe0", "vfe0_src", "vfe0_axi"},
+		.clock_rate = { { 0 },
+				{ 380000000, 510000000, 637000000, 760000000 },
+				{ 0 } },
+		.reg = { "vfe0" },
+		.interrupt = { "vfe0" },
+		.vfe = {
+			.line_num = 4,
+			.has_pd = true,
+			.pd_name = "ife0",
+			.hw_ops = &vfe_ops_170, /* 175 */
+			.formats_rdi = &vfe_formats_rdi_845,
+			.formats_pix = &vfe_formats_pix_845
+		}
+	},
+	/* VFE1 */
+	{
+		.regulators = {},
+		.clock = { "vfe1", "vfe1_src", "vfe1_axi"},
+		.clock_rate = { { 0 },
+				{ 380000000, 510000000, 637000000, 760000000 },
+				{ 0 } },
+		.reg = { "vfe1" },
+		.interrupt = { "vfe1" },
+		.vfe = {
+			.line_num = 4,
+			.has_pd = true,
+			.pd_name = "ife1",
+			.hw_ops = &vfe_ops_170, /* 175 */
+			.formats_rdi = &vfe_formats_rdi_845,
+			.formats_pix = &vfe_formats_pix_845
+		}
+	},
+	/* VFE2 (lite) */
+	{
+		.regulators = {},
+		.clock = { "vfe_lite", "vfe_lite_src", },
+		.clock_rate = { { 320000000, 400000000, 480000000, 600000000 },
+				{ 0 } },
+		.reg = { "vfe_lite" },
+		.interrupt = { "vfe_lite" },
+		.vfe = {
+			.is_lite = true,
+			.line_num = 4,
+			.hw_ops = &vfe_ops_170, /* 175 */
+			.formats_rdi = &vfe_formats_rdi_845,
+			.formats_pix = &vfe_formats_pix_845
+		}
+	},
+};
+
+static const struct resources_icc icc_res_sm7150[] = {
+	{
+		.name = "cam_ahb",
+		.icc_bw_tbl.avg = 60000,
+		.icc_bw_tbl.peak = 120000,
+	},
+	{
+		.name = "cam_hf_0_mnoc",
+		.icc_bw_tbl.avg = 2097152,
+		.icc_bw_tbl.peak = 2097152,
+	},
+	{
+		.name = "cam_sf_0_mnoc",
+		.icc_bw_tbl.avg = 0,
+		.icc_bw_tbl.peak = 2097152,
+	},
+	{
+		.name = "cam_sf_icp_mnoc",
+		.icc_bw_tbl.avg = 2097152,
+		.icc_bw_tbl.peak = 2097152,
+	},
+};
+
 static const struct camss_subdev_resources csiphy_res_8250[] = {
 	/* CSIPHY0 */
 	{
@@ -5655,6 +5843,20 @@ static const struct camss_resources sm6350_resources = {
 	.vfe_num = ARRAY_SIZE(vfe_res_sm6350),
 };
 
+static const struct camss_resources sm7150_resources = {
+	.version = CAMSS_7150,
+	.pd_name = "top",
+	.csiphy_res = csiphy_res_7150,
+	.csid_res = csid_res_7150,
+	.vfe_res = vfe_res_7150,
+	.icc_res = icc_res_sm7150,
+	.icc_path_num = ARRAY_SIZE(icc_res_sm7150),
+	.csiphy_num = ARRAY_SIZE(csiphy_res_7150),
+	.csid_num = ARRAY_SIZE(csid_res_7150),
+	.vfe_num = ARRAY_SIZE(vfe_res_7150),
+	.link_entities = camss_link_entities
+};
+
 static const struct camss_resources sm8250_resources = {
 	.version = CAMSS_8250,
 	.pd_name = "top",
@@ -5754,6 +5956,7 @@ static const struct of_device_id camss_dt_match[] = {
 	{ .compatible = "qcom,sdm845-camss", .data = &sdm845_resources },
 	{ .compatible = "qcom,sm6150-camss", .data = &sm6150_resources },
 	{ .compatible = "qcom,sm6350-camss", .data = &sm6350_resources },
+	{ .compatible = "qcom,sm7150-camss", .data = &sm7150_resources },
 	{ .compatible = "qcom,sm8250-camss", .data = &sm8250_resources },
 	{ .compatible = "qcom,sm8550-camss", .data = &sm8550_resources },
 	{ .compatible = "qcom,sm8650-camss", .data = &sm8650_resources },
-- 
2.43.0

